cordova-plugin-outline/android/java/org/outline/OutlinePlugin.java

```diff
@@ -114,64 +113,78 @@ public boolean is(final String action) {
     }
   }
 
-  // Extra parameters for Intent broadcasting.
-  public enum IntentExtra {
-    TUNNEL_ID("tunnelIdExtra"),
-    PAYLOAD("payloadExtra"),
-    ERROR_CODE("errorCodeExtra");
+  // IPC message and intent parameters.
+  public enum MessageData {
+    TUNNEL_ID("tunnelId"),
+    TUNNEL_CONFIG("tunnelConfig"),
+    ACTION("action"),
+    PAYLOAD("payload"),
+    ERROR_REPORTING_API_KEY("errorReportingApiKey");
 
     public final String value;
-    IntentExtra(final String value) {
+    MessageData(final String value) {
       this.value = value;
     }
   }
 
+  // Encapsulates parameters to start the VPN asynchronously after requesting user permission.
+  private static class StartVpnRequest {
+    public final JSONArray args;
+    public final CallbackContext callback;
+    public StartVpnRequest(JSONArray args, CallbackContext callback) {
+      this.args = args;
+      this.callback = callback;
+    }
+  }
+
   private static final int REQUEST_CODE_PREPARE_VPN = 100;
-  private static final int RESULT_OK = -1; // Standard activity result: operation succeeded.
-  private static final HashSet<String> TUNNEL_INSTANCE_ACTIONS =
-      new HashSet<String>(Arrays.asList(Action.START.value, Action.STOP.value,
-          Action.IS_RUNNING.value, Action.ON_STATUS_CHANGE.value, Action.IS_REACHABLE.value));
-
-  private VpnTunnelService vpnTunnelService = null;
-  private String startRequestTunnelId = null;
-  private JSONObject startRequestConfig = null;
-  private Map<Pair<String, String>, CallbackContext> listeners = new ConcurrentHashMap();
-
-  // Class to bind to VpnTunnelService.
-  private ServiceConnection serviceConnection =
-      new ServiceConnection() {
-        @Override
-        public void onServiceConnected(ComponentName className, IBinder binder) {
-          vpnTunnelService = ((VpnTunnelService.LocalBinder) binder).getService();
-        }
 
-        @Override
-        public void onServiceDisconnected(ComponentName className) {
-          vpnTunnelService = null;
-        }
-      };
+  private IVpnTunnelService vpnTunnelService;
```

Comment:
It seems like you're deliberately allowing some data races on `vpnTunnelService` and `initErrorReporting`, relying on broad `catch` blocks to catch null pointer exceptions, etc.  I'm OK with that, but I think there should at least be a comment to explain what the threading assumptions are.

You might also want to look at the [Android thread annotations](https://developer.android.com/studio/write/annotations#thread-annotations) to see if that could help clarify some of the threading assumptions in OutlinePlugin and VpnTunnelService.

Reply:
I am no longer setting `vpnTunnelService = null` to avoid the null pointer exception and instead rely on `RemoteException` for clarity. I added a comment on the threading assumptions, but didn't really find the annotations useful - perhaps you have a concrete annotation suggestion?

Reply:
I was thinking of using `@MainThread` and `@WorkerThread` annotations to distinguish methods that run on each thread in OutlinePlugin, but it's not important.  I think the comment you added makes the design nice and clear.